asm/exceptasm.s

```diff
@@ -69,646 +81,891 @@ glabel __osIntTable
-/* 004B58 80003F58 00000000 */   nop   
-/* 004B5C 80003F5C 8D2D000C */  lw    $t5, 0xc($t1)
-/* 004B60 80003F60 01AB6821 */  addu  $t5, $t5, $t3
-/* 004B64 80003F64 01AC001A */  div   $zero, $t5, $t4
-/* 004B68 80003F68 15800002 */  bnez  $t4, .L80003F74
-/* 004B6C 80003F6C 00000000 */   nop   
-/* 004B70 80003F70 0007000D */  break 7
-.L80003F74:
-/* 004B74 80003F74 2401FFFF */  li    $at, -1
-/* 004B78 80003F78 15810004 */  bne   $t4, $at, .L80003F8C
-/* 004B7C 80003F7C 3C018000 */   lui   $at, 0x8000
-/* 004B80 80003F80 15A10002 */  bne   $t5, $at, .L80003F8C
-/* 004B84 80003F84 00000000 */   nop   
-/* 004B88 80003F88 0006000D */  break 6
-.L80003F8C:
-/* 004B8C 80003F8C 8D2C0014 */  lw    $t4, 0x14($t1)
-/* 004B90 80003F90 00006810 */  mfhi  $t5
-/* 004B94 80003F94 000D6880 */  sll   $t5, $t5, 2
-/* 004B98 80003F98 018D6021 */  addu  $t4, $t4, $t5
-/* 004B9C 80003F9C 8D4D0004 */  lw    $t5, 4($t2)
-/* 004BA0 80003FA0 256A0001 */  addiu $t2, $t3, 1
-/* 004BA4 80003FA4 AD8D0000 */  sw    $t5, ($t4)
-/* 004BA8 80003FA8 AD2A0008 */  sw    $t2, 8($t1)
-/* 004BAC 80003FAC 8D2A0000 */  lw    $t2, ($t1)
-/* 004BB0 80003FB0 8D4B0000 */  lw    $t3, ($t2)
-/* 004BB4 80003FB4 11600008 */  beqz  $t3, send_done
-/* 004BB8 80003FB8 00000000 */   nop   
-/* 004BBC 80003FBC 0C001057 */  jal   __osPopThread
-/* 004BC0 80003FC0 01202025 */   move  $a0, $t1
-/* 004BC4 80003FC4 00405025 */  move  $t2, $v0
-/* 004BC8 80003FC8 3C048001 */  lui   $a0, %hi(__osRunQueue) # $a0, 0x8001
-/* 004BCC 80003FCC 01402825 */  move  $a1, $t2
-/* 004BD0 80003FD0 0C001045 */  jal   __osEnqueueThread
-/* 004BD4 80003FD4 2484AD48 */   addiu $a0, %lo(__osRunQueue) # addiu $a0, $a0, -0x52b8
+    # Mark thread as having faulted
+    lui     $at, %hi(__osFaultedThread)
+    sw      $k0, %lo(__osFaultedThread)($at)
+    li      $t1, OS_STATE_STOPPED
+    sh      $t1, THREAD_STATE($k0)
+    li      $t1, OS_FLAG_FAULT
+    sh      $t1, THREAD_FLAGS($k0)
+    # Save C0_BADVADDR
+    mfc0    $t2, C0_BADVADDR
+    sw      $t2, THREAD_BADVADDR($k0)
+    # Post the fault message
+    jal     send_mesg
+     li     $a0, OS_EVENT_FAULT*8
+    # Dispatch next thread
+    j       __osDispatchThread
+     nop
+
+/**
+ *  Handles posting event messages to the listening message queue, if there is one
+ */
+send_mesg:
+    # Load pointer to listening message queue
+    lui     $t2, %hi(__osEventStateTab)
+    addiu   $t2, %lo(__osEventStateTab)
+    addu    $t2, $t2, $a0
+    lw      $t1, ($t2)
+    # Save return address
+    move    $s2, $ra
+    # If there is no listening message queue, done
+    beqz    $t1, send_done
+     nop
+    # Test if the message queue is full, if so don't post the message
+    lw      $t3, MQ_VALIDCOUNT($t1)
+    lw      $t4, MQ_MSGCOUNT($t1)
+    slt     $at, $t3, $t4
+    beqz    $at, send_done
+     nop
+    # Add validcount to first and modulo with msgcount
+    lw      $t5, MQ_FIRST($t1)
+    addu    $t5, $t5, $t3
+    # Modulo
+    div     $zero, $t5, $t4
```

Comment:
how does this `div` work, the various references I'm looking at show it used as `div a, b` and accomplishing `a = b * lo + hi`

Reply:
Here it's used to calculate the remainder rather than the quotient which is more common, lower down it does `mfhi` rather than `mflo`. So your source is correct `a = b * lo + hi` is what `div` does, `lo` and `hi` are of course the quotient and remainder and are extracted via `mflo` and `mfhi` respectively.

Reply:
but why is it `div     $zero, $t5, $t4` instead of idk maybe `div     $t5, $t4`

Reply:
Oh right - This is just how it was originally disassembled, for some reason the gnu assembler accepts 3 args to div, I don't think the first really does anything. I can see if removing it still works since it should.

Reply:
nvm, I found out both why it was disassembled this way and what it does. It's a gnu assembler thing for preventing generating div0 and overflow checks. As you can see around this code, there are checks for div0 and overflow, these were automatically generated by the sgi assembler. The gnu assembler actually does something similar (but it does not match), but if you add the leading `$zero` argument it prevents the assembler from adding the div0 and overflow checks.
So it has to stay this way as long as we use the gnu assembler.